src/helpers/buffers/ordering_sender.rs

```diff
@@ -0,0 +1,566 @@
+#![allow(dead_code)] // TODO remove
+
+use crate::helpers::messaging::Message;
+use futures::{task::Waker, Future, Stream};
+use generic_array::GenericArray;
+use std::{
+    cmp::Ordering,
+    collections::VecDeque,
+    mem::drop,
+    num::NonZeroUsize,
+    ops::Deref,
+    pin::Pin,
+    sync::{
+        atomic::{
+            AtomicUsize,
+            Ordering::{AcqRel, Acquire},
+        },
+        Mutex, MutexGuard,
+    },
+    task::{Context, Poll},
+};
+use typenum::Unsigned;
+
+/// The operating state for an `OrderingSender`.
+struct State {
+    /// A store of bytes to write into.
+    buf: Vec<u8>,
+    /// How many bytes have been written and are available.
+    written: usize,
+
+    /// The sender is closed.
+    closed: bool,
+
+    /// An entity to wake when the buffer is read from.
+    write_ready: Option<Waker>,
+    /// Another entity to wake when the buffer is read from.
+    stream_ready: Option<Waker>,
+}
+
+impl State {
+    fn new(capacity: NonZeroUsize) -> Self {
+        Self {
+            buf: vec![0; capacity.get()],
+            written: 0,
+            closed: false,
+            write_ready: None,
+            stream_ready: None,
+        }
+    }
+
+    fn save_waker(v: &mut Option<Waker>, cx: &Context<'_>) {
+        if let Some(w) = v.replace(cx.waker().clone()) {
+            assert!(cx.waker().will_wake(&w));
+        }
+    }
+
+    fn wake(v: &mut Option<Waker>) {
+        if let Some(w) = v.take() {
+            w.wake();
+        }
+    }
+
+    fn write<M: Message>(&mut self, m: &M, cx: &Context<'_>) -> bool {
+        let b = &mut self.buf[self.written..];
+        if M::Size::USIZE <= b.len() {
+            self.written += M::Size::USIZE;
+            m.serialize(GenericArray::from_mut_slice(&mut b[..M::Size::USIZE]));
+
+            if self.written * 2 >= self.buf.len() {
+                Self::wake(&mut self.stream_ready);
+            }
+            true
+        } else {
+            Self::save_waker(&mut self.write_ready, cx);
```

Comment:
maybe worth doing `debug_assert(M::Size::USIZE <= b.capacity())` to better handle obvious cases where progress cannot be made at all?